### PR TITLE
SystemTextJson: Add option to reject null string values

### PR DESCRIPTION
**Pull request by nordfjord**
Adds a `StringConverter` to the converter list by default. This converter will throw when deserializing a string that's set to `null`

Before this change this would return null:

```fsharp
serde.Deserialize<string>("null")
```

After this change the above throws. Also affects properties on records.

---

I'm also updating FsCheck to the 3.0.0-* branch because it doesn't emit null as a legal string

### CODE REVIEW

**Comment by bartelink** [comment]
TL;DR if this was off by default, I'd probably just merge it without much discussion. If I truly hated it as a default, I'd insist on it being changed. However, the truth is I'm torn.
----

In general I agree that apps are better off with `null`-valued Strings not getting into serialized data

I don't believe anything in github.com/jet will break based on it (but I had to check - e.g. the `correlationId` and `causationId` values from IEventData may be `null`, but https://github.com/jet/equinox/blob/master/src/Equinox.CosmosStore/CosmosStore.fs#L1200 will permit that

The elephant in the room is the fact that this is a breaking change for FsCodec an'd we're not in version 0.x. As we're in a phase where FsCodec and Equinox are in RC, that's less significant than it ordinarily would be (and it's not a binary breaking change). However, I can definitely picture people being surprised and ending up with broken production systems.

The simplest way to address this is to add an optional argument to `Options.Create` which lets one opt into this.

The next level is to make it a non-optional argument in `Options.Create` to force a decision. That's less desirable though
- the `Codec` types are defined as (and implemented as) defaulting to the same settings as `Options.Create()`
- atm there's an ability to create a serialization profile by via the pretty terse `let serdes = Serdes Options.Default` as mentioned in https://github.com/jet/fscodec#examples-of-using-serdes-to-define-a-contract - if you add an argument, those all get messier.

As it stands, this is a thus a pretty big change to force on people.

Other aspects:
- The intention is for FsCodec to provide an equivalent experience OOTB in JSON.net and STJ to the max degree possible (e.g. Options are handled, DUs are not). Atm off the top of my head the notable diffs in the experience are: 
  1. JSON.NET has the ugly DU support by default, STJ throws (ideally we'd probably throw - plugging that hole on the Newtonsoft side would be a bugfix to my mind)
  2. JSON.NET has broken `list` support, STJ just works (not sure about Map, but its advised against in the README)
  3. FsCodec.NewtonsoftJson has an opt in to exclude empty fields, it's not implemented on the STJ side
  4. On the STJ side, there's opt in support for default TypeSafeEnum encoding and the safe DU encoding (but that could be ported)
- While the main principle is to give no surprises to the average dev that has used STJ and/or Newtonsoft _but is not the unfortunately that has spent days writing the ugly serialization helpers in a medium to large size prod system and has hence been forced to know details on this sort of thing_, there is decent docs coverage and someone scanning the README should come away understanding the key diffs between FsCodec and vanilla STJ/JSON.NET.

In my world, I tend to have two places in my app where I would be able to configure things (and I personally would immediately switch it on):
- <App.>Serdes (a default instance that's used)
- <App.>EventCodec.gen 
- The first one would then also flow through to my default ASP.NET config by borrowing the converter via https://github.com/jet/fscodec#aspnet-core-with-systemtextjson

Thoughts, @deviousasti @enricosada @wantastic84 ?
----
Conclusions:
1. I do want this converter in the lib
2. Whatever it is that's concluded, it needs to be well-documented in the xmldoc and in the readme

These are the doc sections this should be covered in (perhaps a dedicated section near the top with an anchor that can be referenced by the sections might help?)
- https://github.com/jet/fscodec#common-api
- https://github.com/jet/fscodec#fscodecsystemtextjsonoptions
- https://github.com/jet/fscodec#systemtextjson-specific-low-level-converters

File: src/FsCodec.SystemTextJson/Options.fs [hunk 1]
```diff
@@ -59,12 +59,14 @@ type Options private () =
             /// <summary>Apply <c>UnionConverter</c> for all Discriminated Unions, if <c>TypeSafeEnumConverter</c> not possible. Defaults to <c>false</c>.</summary>
             [<Optional; DefaultParameterValue(null)>] ?autoUnionToJsonObject : bool) =
 
+        let defaultConverters: JsonConverter array  = [| StringConverter() |]
```
Comment (bartelink) [inline src/FsCodec.SystemTextJson/Options.fs:62]: there needs to be a way to opt out. See top level comment re considerations for defaulting to having it on, forcing a decision, or defaulting to off.
Whatever is done in the API here also needs to happen on the STJ side (unless it defaults to off, in which case it can be a new option that's just not been ported _yet_)

File: src/FsCodec.SystemTextJson/StringConverter.fs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+type StringConverter() =
+    inherit JsonConverter<string>()
+
+    override _.HandleNull = true
+    override _.CanConvert(t) = t = typeof<string>
+
+    override this.Read(reader, _typeToConvert, _options) =
+        let value = reader.GetString()
+        if value = null then failwith "Expected string, got null."
```
Comment (bartelink) [inline src/FsCodec.SystemTextJson/StringConverter.fs:13]: message is fine, but probably invalidOp, nullArg, ArgumentOutOfRangeException or someting (not sure on prior art for a good exception type for this in the context of a JSON lib, maybe FSharp.STJ has picked an approach?)

File: tests/FsCodec.NewtonsoftJson.Tests/SomeNullHandlingTests.fs [hunk 3]
```diff
@@ -9,10 +9,7 @@ let serdes = Serdes Options.Default
 
 let [<Fact>] ``Options.Create does not roundtrip Some null`` () =
     let value : string option = Some null
-    let ser = serdes.Serialize value
-    "null" =! ser
-    // But it doesn't roundtrip
-    value <>! serdes.Deserialize ser
+    raises <@ serdes.Serialize value @>
```
Comment (bartelink) [inline tests/FsCodec.NewtonsoftJson.Tests/SomeNullHandlingTests.fs:12]: Probably in SerdesTests.fs, would be good to have characterization tests that illustrate the default config for STJ and NSJ is to silently do roundtripping of `null` (and the fact that `Serdes <| Options.CreateDefault()` preserves that), see 
 `type StjCharacterization`

File: tests/FsCodec.NewtonsoftJson.Tests/UnionConverterTests.fs [hunk 4]
```diff
@@ -7,11 +7,12 @@ open System.Text.Json.Serialization
 #else
 module FsCodec.NewtonsoftJson.Tests.UnionConverterTests
 
+open FsCheck.FSharp
```
Comment (bartelink) [inline tests/FsCodec.NewtonsoftJson.Tests/UnionConverterTests.fs:10]: alreayd on L15

File: tests/FsCodec.NewtonsoftJson.Tests/UnionConverterTests.fs [hunk 5]
```diff
@@ -151,8 +152,8 @@ let ``deserializes properly`` () =
     test <@ CaseE ("hi", 0) = deserialize """{"case":"CaseE","Item1":"hi","Item2":0}""" @>
 #if !SYSTEM_TEXT_JSON
     // NB this only passes by virtue of MissingMemberHandling=Ignore and NullValueHandling=Ignore in default settings
-#endif
     test <@ CaseE (null, 0) = deserialize """{"case":"CaseE","Item1":null,"Item2":0,"Item3":"hi","Item4":"0"}""" @>
```
Comment (bartelink) [inline tests/FsCodec.NewtonsoftJson.Tests/UnionConverterTests.fs:155]: needs comment a to reasoning

File: tests/FsCodec.SystemTextJson.Tests/AutoUnionTests.fs [hunk 6]
```diff
@@ -22,6 +22,10 @@ let [<Xunit.Fact>] ``Basic characteristics`` () =
     test <@ Not { body = "A"; opt = None; list = [] } = serdes.Deserialize "{\"case\":\"Not\",\"body\":\"A\",\"list\":[]}" @>
     test <@ Not { body = "A"; opt = None; list = ["A"] } = serdes.Deserialize "{\"case\":\"Not\",\"body\":\"A\",\"list\":[\"A\"]}" @>
 
+let [<Xunit.Fact>] ``It throws on null strings`` () =
+    raises <@ serdes.Deserialize<string> "null" @>
+    raises <@ serdes.Deserialize<NotAUnion> "{\"body\": null, \"opt\": null, \"list\": []}" @>
```
Comment (bartelink) [inline tests/FsCodec.SystemTextJson.Tests/AutoUnionTests.fs:27]: How can one roundtrip and/or safely handle `null` values if this throws?
Need to have a way to handle `null` values in a blob of JSON without having to drop to low level STJ overrides

File: tests/FsCodec.SystemTextJson.Tests/AutoUnionTests.fs [hunk 7]
```diff
@@ -22,6 +22,10 @@ let [<Xunit.Fact>] ``Basic characteristics`` () =
     test <@ Not { body = "A"; opt = None; list = [] } = serdes.Deserialize "{\"case\":\"Not\",\"body\":\"A\",\"list\":[]}" @>
     test <@ Not { body = "A"; opt = None; list = ["A"] } = serdes.Deserialize "{\"case\":\"Not\",\"body\":\"A\",\"list\":[\"A\"]}" @>
 
+let [<Xunit.Fact>] ``It throws on null strings`` () =
+    raises <@ serdes.Deserialize<string> "null" @>
```
Comment (bartelink) [inline tests/FsCodec.SystemTextJson.Tests/AutoUnionTests.fs:26]: needs to go in the low level characterisation tests e.g. somewhere in `SedesTests.fs` and/or `StjCharacterization`

File: tests/FsCodec.SystemTextJson.Tests/AutoUnionTests.fs (hunk 6 above)
Comment (nordfjord) [reply tests/FsCodec.SystemTextJson.Tests/AutoUnionTests.fs:27]: Options are supported. This forces you to make the nullability explicit. Though you might be asking for something different?

Comment (bartelink) [reply tests/FsCodec.SystemTextJson.Tests/AutoUnionTests.fs:27]: ok but given opt is a `string option`, I would expect the `null` to then come back as either `None` or `Some null`, whereas this suggests it will throw when presented with that ?

More importantly, if I had emitted a `None` via `string option`, would I be able to load it from whatever that would have rendered as ?

Comment (nordfjord) [reply tests/FsCodec.SystemTextJson.Tests/AutoUnionTests.fs:27]: Ah, the error is because the body is null. The opt deserializes fine

Comment (bartelink) [reply tests/FsCodec.SystemTextJson.Tests/AutoUnionTests.fs:27]: Ah, that makes sense. (Though I guess that's a hint that having a specifically typed exception with a message that pinpoints the issue to `null` not being allowed for `string` _unless wrapped in an Option_ will make this the best experience for when people inevitably run into it.)

File: src/FsCodec.SystemTextJson/Options.fs (hunk 1 above)
Comment (nordfjord) [reply src/FsCodec.SystemTextJson/Options.fs]: I've made it optional for now

File: tests/FsCodec.SystemTextJson.Tests/AutoUnionTests.fs (hunk 6 above)
Comment (nordfjord) [reply tests/FsCodec.SystemTextJson.Tests/AutoUnionTests.fs]: I've changed the message to indicate what needs to be done to remedy it.

File: src/FsCodec.SystemTextJson/Options.fs [hunk 8]
```diff
@@ -57,14 +57,19 @@ type Options private () =
             /// <summary>Apply <c>TypeSafeEnumConverter</c> if possible. Defaults to <c>false</c>.</summary>
             [<Optional; DefaultParameterValue(null)>] ?autoTypeSafeEnumToJsonString : bool,
             /// <summary>Apply <c>UnionConverter</c> for all Discriminated Unions, if <c>TypeSafeEnumConverter</c> not possible. Defaults to <c>false</c>.</summary>
-            [<Optional; DefaultParameterValue(null)>] ?autoUnionToJsonObject : bool) =
+            [<Optional; DefaultParameterValue(null)>] ?autoUnionToJsonObject : bool,
+            /// <summary>When set to <c>false</c> the codec will throw on <c>null</c> strings. Use <c>string option</c> to allow nulls.
+            [<Optional; DefaultParameterValue(null)>] ?allowNullStrings : bool) =
+
+        let defaultConverters: JsonConverter array =
+            if allowNullStrings = Some false then [| RejectNullStringConverter() |] else Array.empty
```
Comment (bartelink) [inline src/FsCodec.SystemTextJson/Options.fs:65]: I know I used this, and find it very legible
But FYI this syntax constructs a FSharpOption
Does not truly matter here but maybe
```suggestion
            if defaultArg allowNullStrings true then [| RejectNullStringConverter() |] else Array.empty
```
(please don't use)

File: src/FsCodec.SystemTextJson/Options.fs [hunk 9]
```diff
@@ -57,14 +57,19 @@ type Options private () =
             /// <summary>Apply <c>TypeSafeEnumConverter</c> if possible. Defaults to <c>false</c>.</summary>
             [<Optional; DefaultParameterValue(null)>] ?autoTypeSafeEnumToJsonString : bool,
             /// <summary>Apply <c>UnionConverter</c> for all Discriminated Unions, if <c>TypeSafeEnumConverter</c> not possible. Defaults to <c>false</c>.</summary>
-            [<Optional; DefaultParameterValue(null)>] ?autoUnionToJsonObject : bool) =
+            [<Optional; DefaultParameterValue(null)>] ?autoUnionToJsonObject : bool,
+            /// <summary>When set to <c>false</c> the codec will throw on <c>null</c> strings. Use <c>string option</c> to allow nulls.
```
Comment (bartelink) [inline src/FsCodec.SystemTextJson/Options.fs:61]: For defaults that default to `true`, and for better alignment with the converter name, suggest flipping the arg name and semantics to `rejectNullStrings` 
```suggestion
            /// <summary>Apply <c>RejectNullStringConverter</c> in order to have serialization throw on <c>null</c> strings. Use <c>string option</c> to represent strings that can potentially be <c>null</c>.
```

File: src/FsCodec.SystemTextJson/Options.fs [hunk 10]
```diff
@@ -57,14 +57,19 @@ type Options private () =
...
+            if allowNullStrings = Some false then [| RejectNullStringConverter() |] else Array.empty
+        let converters = if converters = null then defaultConverters else Array.append converters defaultConverters
 
         Options.CreateDefault(
             converters =
                 (   match autoTypeSafeEnumToJsonString = Some true, autoUnionToJsonObject = Some true with
                     | tse, u when tse || u ->
-                        let converter : JsonConverter array = [| UnionOrTypeSafeEnumConverterFactory(typeSafeEnum = tse, union = u) |]
-                        if converters = null then converter else Array.append converters converter
+                        Array.append converters [| UnionOrTypeSafeEnumConverterFactory(typeSafeEnum = tse, union = u) |]
```
Comment (bartelink) [inline src/FsCodec.SystemTextJson/Options.fs:72]: Now there are >=2 conditions, it might make sense to change the converter list to an array comprehension

File: src/FsCodec.SystemTextJson/RejectNullStringConverter.fs [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+type RejectNullStringConverter() =
+    inherit JsonConverter<string>()
+
+    override _.HandleNull = true
+    override _.CanConvert(t) = t = typeof<string>
+
+    override this.Read(reader, _typeToConvert, _options) =
+        let value = reader.GetString()
+        if value = null then nullArg "Expected string, got null. When allowNullStrings is false you must explicitly type optional strings as 'string option'"
```
Comment (bartelink) [inline src/FsCodec.SystemTextJson/RejectNullStringConverter.fs:13]: maybe add a `[<Literal>]`, maybe a function to keep the message DRY?

File: tests/FsCodec.SystemTextJson.Tests/SerdesTests.fs [hunk 12]
```diff
@@ -121,3 +122,27 @@ let [<Fact>] ``Switches off the HTML over-escaping mechanism`` () =
     test <@ ser = """{"a":1,"b":"\"+"}""" @>
     let des = serdes.Deserialize ser
     test <@ value = des @>
+
+let [<Fact>] ``Rejects null strings`` () =
+    let serdes = Serdes(Options.Create(allowNullStrings = false))
+    let value: string = null
+    raises <@ serdes.Serialize value @>
```
Comment (bartelink) [inline tests/FsCodec.SystemTextJson.Tests/SerdesTests.fs:129]: are they all ANE ? might be worth doing raisesWith to pin that?

File: tests/FsCodec.SystemTextJson.Tests/SerdesTests.fs [hunk 13]
```diff
@@ -121,3 +122,27 @@ let [<Fact>] ``Switches off the HTML over-escaping mechanism`` () =
...
+
+    let value = [| "A"; null |]
+    raises <@ serdes.Serialize value @>
+
+    let value = [| Some "A"; None |]
```
[the remark this answers is not in the document]
Comment (bartelink) [reply tests/FsCodec.SystemTextJson.Tests/SerdesTests.fs:134]: (and if the point is that this works/needs to work with and without the converter, I would maybe make ti a Theory with true and false args being paseed to Options.Create)

File: src/FsCodec.SystemTextJson/Options.fs (hunk 1 above)
Comment (bartelink) [reply src/FsCodec.SystemTextJson/Options.fs]: I guess that's all good then!

File: tests/FsCodec.NewtonsoftJson.Tests/SomeNullHandlingTests.fs (hunk 3 above)
Comment (bartelink) [reply tests/FsCodec.NewtonsoftJson.Tests/SomeNullHandlingTests.fs]: Resolving, see comment about illustrating that some properties hold with and without the rejection converter in other places

**Comment by bartelink** [comment]
OK now that the default is flipped, the main remaining thing other than the minor review comment things is to mention the converter in the README in the cited places (perhaps with a small section illustrating how `string option` can be used in a messy case where the input is conditional in some input?)

Am still slightly torn on what the default should be, but I guess the simplification of the PR shows its probably for the best...

**Comment by bartelink** [comment]
Looks great, thanks!
@enricosada @deviousasti @wantastic84 thoughts on whether this should be a default?
